src/lib/coreaudio/CoreAudioHelpers.cxx

```diff
@@ -0,0 +1,261 @@
+				<< "]";
+			break;
+	}
+	return str.c_str();
+}
+
+AudioBufferList *
+AllocateABL(const AudioStreamBasicDescription asbd, const UInt32 capacity_frames)
+{
+	AudioBufferList *buffer_list = nullptr;
+	UInt32 num_buffers = (asbd.mFormatFlags & kAudioFormatFlagIsNonInterleaved) ? asbd.mChannelsPerFrame : 1;
+	
+	if((buffer_list = static_cast<AudioBufferList *>(calloc(1, offsetof(AudioBufferList, mBuffers) + (sizeof(AudioBuffer) * num_buffers)))) == nullptr)
+		throw std::runtime_error("Unable to allocate memory for AudioBufferList.");
+	else
+	{
+		buffer_list->mNumberBuffers = num_buffers;
+		for(UInt32 buffer_index = 0; buffer_index < buffer_list->mNumberBuffers; ++buffer_index)
+		{
+			try
+			{
+				AllocateAudioBuffer(buffer_list->mBuffers[buffer_index], asbd, capacity_frames);
+			}
+			catch (...)
+			{
+				DeallocateABL(buffer_list);
+				std::throw_with_nested("Unable to allocate memory for AudioBufferList.");
+			}
+		}
+	}
+	return buffer_list;
+}
+
+void
+DeallocateABL(AudioBufferList *buffer_list)
+{
+	if(buffer_list != nullptr)
+	{
+		for(UInt32 buffer_index = 0; buffer_index < buffer_list->mNumberBuffers; ++buffer_index) {
+			free(buffer_list->mBuffers[buffer_index].mData);
+		}
+	}
+	free(buffer_list);
+}
+
+void
+AllocateAudioBuffer(AudioBuffer &buffer, const AudioStreamBasicDescription asbd, const UInt32 capacity_frames)
+{
+	if((buffer.mData = static_cast<void *>(calloc(capacity_frames, asbd.mBytesPerFrame))) == nullptr)
+	   throw std::runtime_error("Unable to allocate memory for AudioBuffer.");
+	buffer.mDataByteSize = capacity_frames * asbd.mBytesPerFrame;
+	buffer.mNumberChannels = (asbd.mFormatFlags & kAudioFormatFlagIsNonInterleaved) ? 1 : asbd.mChannelsPerFrame;
+}
+
+AudioStreamBasicDescription
+AudioFormatToASBD(AudioFormat format)
+{
+	assert(format.format != SampleFormat::UNDEFINED);
+	assert(format.format != SampleFormat::DSD);
+	
+	AudioStreamBasicDescription out_format;
+	memset(&out_format, 0, sizeof(out_format));
+	out_format.mSampleRate = format.sample_rate;
+	out_format.mChannelsPerFrame = format.channels;
+	out_format.mFormatID = kAudioFormatLinearPCM;
+	out_format.mFramesPerPacket = 1;
+	out_format.mBytesPerPacket = out_format.mBytesPerFrame = format.GetFrameSize();
+	
+	switch (format.format)
+	{
+		case SampleFormat::S8:
+			out_format.mBitsPerChannel = 8;
+			out_format.mFormatFlags = kAudioFormatFlagIsSignedInteger;
+			break;
+		case SampleFormat::S16:
+			out_format.mBitsPerChannel = 16;
+			out_format.mFormatFlags = kAudioFormatFlagIsSignedInteger;
+			break;
+		case SampleFormat::S24_P32:
+			out_format.mBitsPerChannel = 24;
+			out_format.mFormatFlags = kAudioFormatFlagIsSignedInteger;
+			break;
+		case SampleFormat::S32:
+			out_format.mBitsPerChannel = 32;
+			out_format.mFormatFlags = kAudioFormatFlagIsSignedInteger;
+			break;
+		case SampleFormat::FLOAT:
+			out_format.mBitsPerChannel = 32;
+			out_format.mFormatFlags = kAudioFormatFlagIsFloat;
+			break;
+		default:
+			break;
```

Comment:
Uninitialized memory in this case.

Reply:
This already considers all possible formats (except UNKNOWN and DSD which will be caught by the assertion). Added gcc_unreachable here.

Reply:
OK, added the two additional elements of the SampleFormat enum